### PR TITLE
Generalize label input sparsity check and refactor

### PR DESCRIPTION
**Pull request by guyang3532**
### Description
The InsertGatherBeforeSceLoss optimization is enabled when the density of label padding less than 90%. We need to check the density of the label padding to decide whether enable the optimization.

Before this pr, we just check the inputs of graph and correlate one with the SCE node by iterate graph from the SCE node back to one graph input.
This is hard to be general because there may be complicated pattern between graph input and SCE node.

This pr check padding density by the direct input of SCE module rather than the input of graph at the first graph execution when exporting onnx graph.
And if the density < 90%, insert a flag PythonOp after the SCE node as:
```
           SoftmaxCrossEntropy
		  |
            PythonOp (func_name: FlagAndPrintDensity)   (insert if density < 90%)
		  |
            Following graph
```

When the InsertGatherBeforeSceLoss is invoked, it check if there is the flag PythonOp(func_name: FlagAndPrintDensity) after the SCE node and if it is, remove it and do the padding elimination optimization.

If the env of ORTMODULE_PRINT_INPUT_DENSITY is 1, we will print input density each step by the PythonOp (func_name: FlagAndPrintDensity). In this case the PythonOp will not be removed.

### CODE REVIEW

File: orttraining/orttraining/core/optimizer/graph_transformer_config.h [hunk 1]
```diff
@@ -25,6 +25,8 @@ struct TrainingGraphTransformerConfiguration : public GraphTransformerConfigurat
   // Enable compute optimizer.
   bool enable_compute_optimizer{false};
 
+  bool print_input_density{false};
+
   // Enable label sparsity compute optimization for the input names in the below list.
   std::vector<std::string> sparse_label_input_names;
```
Comment (pengwa) [inline orttraining/orttraining/core/optimizer/graph_transformer_config.h:31]: remove 'sparse_label_input_names'?

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 2]
```diff
@@ -730,11 +722,47 @@ def _embedding_hook(module, args, output):
...
+        label_check_hook_handles = []
+        for name, sub_module in self._flattened_module.named_modules():
+            if isinstance(sub_module, torch.nn.modules.loss.CrossEntropyLoss):
+                label_check_hook_handles.append(sub_module.register_forward_pre_hook(partial(_label_hook, name)))
+
+        return label_check_hook_handles
+
     @_logger.TrackTime(_logger.ORTModuleInitPhase.DETECTION)
     def _enable_conditional_optimizations(
         self, graph_transformer_config: C.TrainingGraphTransformerConfiguration, inputs: Tuple, kwargs: Dict
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:768]:
```suggestion
        self, inputs: Tuple, kwargs: Dict
```

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 3]
```diff
@@ -730,11 +722,47 @@ def _embedding_hook(module, args, output):
...
+
+        label_check_hook_handles = []
+        for name, sub_module in self._flattened_module.named_modules():
+            if isinstance(sub_module, torch.nn.modules.loss.CrossEntropyLoss):
+                label_check_hook_handles.append(sub_module.register_forward_pre_hook(partial(_label_hook, name)))
+
+        return label_check_hook_handles
+
     @_logger.TrackTime(_logger.ORTModuleInitPhase.DETECTION)
     def _enable_conditional_optimizations(
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:767]:
```suggestion
    def _detect_inputs(
```

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 4]
```diff
@@ -730,11 +722,47 @@ def _embedding_hook(module, args, output):
...
+                    embedding_hook_handles.append(sub_module.register_forward_pre_hook(partial(_embedding_hook, name)))
 
         return embedding_hook_handles
 
+    def _add_check_label_sparsity_hook(self):
+        """
+        Add hook to check label sparsity and enable sceloss compute optimization if applicable.
+        1. Register forward pre hook to the sceloss module in the model and the hook will check sparsity of the label input.
+        2. If the sparsity is below a threshold, enable sceloss compute optimization by adding FlagSceLossSparse after the
+           output. GraphTransformer of InsertGatherBeforeSceLoss will check the FlagSceLossSparse and do the actual
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:734]:
```suggestion
           output. GraphTransformer of InsertGatherBeforeSceLoss will check the FlagAndPrintDensity and do the actual
```

File: orttraining/orttraining/core/optimizer/graph_transformer_config.h (hunk 1 above)
Comment (guyang3532) [reply orttraining/orttraining/core/optimizer/graph_transformer_config.h]: removed, thx.

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 2 above)
Comment (guyang3532) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py]: fixed, thx

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 4 above)
Comment (guyang3532) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py]: fixed, thx

**Review by pengwa — APPROVED** [review]
